### PR TITLE
feat: add 10 React Native & Expo diagnostic rules

### PR DESCRIPTION
**Pull request by rayhanadev**
## What

Adds 10 net-new React Native & Expo diagnostics, researched against first-party docs/RFCs (RN docs, RFC 0894/0759, New-Arch WG, Expo docs, eslint-config-expo, Callstack optimization guide, rn-diff-purge) and **validated against an OSS corpus** before landing. Six are oxlint AST rules; four are project-level checks gated on the RN/Expo capability and run in the environment-checks phase (skipped in diff/staged mode), alongside the existing `checkExpoProject` / `checkPnpmHardening`.

### AST rules (`packages/oxlint-plugin-react-doctor`)

| Rule | Sev | Catches |
|---|---|---|
| `rn-no-deep-imports` | warn | Deep imports of public symbols from `react-native/Libraries/*` (RFC 0894; breaks on upgrade). Curated to root-re-exported symbols + tailored `NewAppScreen` message; skips type-only imports and the Codegen/TurboModule surface. |
| `rn-no-set-native-props` | warn | `ref.current(?.).setNativeProps(...)` — a silent no-op under the New Architecture (Fabric). |
| `rn-no-image-children` | error | Children inside react-native `<Image>` (renders none) → `<ImageBackground>`. Import-resolved, so `expo-image`/custom `Image` are ignored. |
| `rn-no-panresponder` | warn | `PanResponder` from `react-native` (JS-thread gestures) → `react-native-gesture-handler`. |
| `rn-detox-missing-await` | warn | Un-awaited Detox actions / `waitFor` / `expect(element(...))` in `*.e2e.*` files. |
| `expo-no-non-inlined-env` | warn | Computed `process.env[...]` + `process.env` destructuring that `babel-preset-expo` can't inline (→ `undefined` at runtime); scoped to Expo client files. |

### Project-level checks (`packages/core`)

| Rule | Catches |
|---|---|
| `rn-no-metro-babel-preset` | `module:metro-react-native-babel-preset` in a babel config (renamed to `@react-native/babel-preset`; uninstalled on RN 0.73+). |
| `rn-library-react-in-dependencies` | A `react-native-builder-bob` library listing `react`/`react-native` in `dependencies` instead of `peerDependencies`. |
| `expo-reanimated-v4-requires-new-arch` | `react-native-reanimated` v4 with `newArchEnabled: false` in the app config (first-launch crash). |
| `expo-updates-no-unsafe-production-config` | `updates.disableAntiBrickingMeasures: true` in the app config (can brick installed apps). |

New `checkReactNativeProject` aggregator (RN/Expo-gated) wired into `run-inspect`; the two Expo checks join `checkExpoProject` and share a new static `app.json`/`app.config.*` reader.

## Eval validation

Scanned the built rules over an OSS RN/Expo corpus (666 rootDirs / 103 repos) with `--warnings`. 82 hits, triaged against source:

- **8/10 clean true positives**; the 3 guardrail rules correctly produced 0 (no corpus project hits them).
- **2 false positives found and fixed with regression tests:** `rn-library-react-in-dependencies` fired on an `example/` app (now gated strictly on the bob config block, not the devDep); `rn-no-deep-imports` fired on `import { linear } from ".../Animated/Easing"` (root exports `Easing`, not `linear`) — dropped the namespace module and made the message generic. Re-scan confirmed both FPs clean with true positives preserved.

Notably, `rn-detox-missing-await` surfaced real un-awaited `waitFor()` calls in `react-native-screens`' own e2e tests.

## Notes

- Detailed research, rule contracts, edge-case audit, and eval results are in `docs/react-native-expo-rule-candidates.md` and `docs/rn-expo-rule-contracts.md`.
- This is the validated Tier-1 batch; ~30 lower-tier candidates remain documented for follow-up.

## Test plan

- [x] `pnpm --filter oxlint-plugin-react-doctor gen` (registry regen)
- [x] Focused unit + regression tests for all 10 rules pass
- [x] `pnpm --filter oxlint-plugin-react-doctor typecheck` / `@react-doctor/core typecheck`
- [x] `pnpm typecheck` (full repo, turbo)
- [x] lint + format clean on changed files
- [x] core suite green except one pre-existing, environment-specific `env-local` git test (global `~/.config/git/ignore` ignores `.env.local`)

Made with [Cursor](https://cursor.com)

---

<a href="https://app.devin.ai/review/millionco/react-doctor/pull/625" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

---

> [!NOTE]
> **Medium Risk**
> New lint and offline config checks can surface widely in RN/Expo repos; severity and file-scoping (e.g. dynamic app config skipped) limit false positives but behavior is still user-visible on full scans.
> 
> **Overview**
> Adds **10 React Native & Expo diagnostics** grounded in first-party docs/RFCs and OSS corpus validation: **six oxlint AST rules** (deep `react-native/Libraries` imports, Fabric-no-op `setNativeProps`, RN `<Image>` children, `PanResponder`, Detox missing `await`, Expo non-inlinable `process.env`) and **four project-level checks** run in the existing **environment-checks** phase (skipped in diff/staged mode).
> 
> Introduces **`checkReactNativeProject`** for RN-wide manifest/config issues (legacy Metro Babel preset, builder-bob libraries putting `react`/`react-native` in `dependencies`) and extends **`checkExpoProject`** with Reanimated v4 + `newArchEnabled: false` and unsafe `expo-updates` anti-bricking flags, backed by a shared static **`readExpoAppConfig`** helper that bails when a dynamic `app.config.*` is present. Rules are registered in the oxlint plugin; research/contracts live in new `docs/` files.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9a5a4b7052efb8b1e9a1e45ba960ab51fce136b1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-no-deep-imports.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+      if (finding) context.report({ node, message: finding.message });
+    };
+
+    return {
+      ImportDeclaration(node: EsTreeNodeOfType<"ImportDeclaration">) {
+        // Skip `import type { ... }` — many RN internal *types* are not
+        // re-exported from the root, so "import from react-native" would
+        // be wrong advice for them.
+        if (node.importKind === "type") return;
+        reportFinding(node, node.source?.value);
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/oxlint-plugin-react-doctor/src/plugin/rules/react-native/rn-no-deep-imports.ts:156]: Fixed in `b9ddf74b`: `rn-no-deep-imports` now also skips inline all-type imports (`import { type X } from "react-native/Libraries/..."`) and re-exports (`export { type X } from`), matching the existing declaration-level `import type` skip — RN internal types often aren't root-exported. A mixed `import { Value, type T }` keeps a value specifier and is still flagged. Regression tests added.

File: packages/core/src/checks/expo/utils/read-expo-app-config.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    let parsed: unknown;
+    try {
+      parsed = JSON.parse(contents);
+    } catch {
+      continue;
+    }
+    const config = unwrapExpoConfig(parsed);
+    if (config) return { config, configFile: fileName };
+  }
+  return { config: null, configFile: null };
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/core/src/checks/expo/utils/read-expo-app-config.ts]: Valid — fixed in `e99d4f22`. `readExpoAppConfig` now detects a dynamic `app.config.{js,ts,cjs,mjs}` and, when one exists, returns no config so both checks bail. Since Expo passes the static config into the dynamic file and uses its return value (which can flip `newArchEnabled`/`disableAntiBrickingMeasures`), trusting `app.json` there risked a false positive; we now treat it as a documented false-negative instead. Regression tests added for both checks (stale app.json value + a dynamic config present → no diagnostic).

File: packages/core/src/checks/expo/utils/read-expo-app-config.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+] as const;
+
+export interface ExpoAppConfig {
+  /** Parsed `expo` config object from a JSON app config, or null. */
+  readonly config: Record<string, unknown> | null;
+  /** The file `config` was parsed from (so checks can report it), or null. */
+  readonly configFile: string | null;
+}
+
+const readFileSafe = (filePath: string): string | null => {
```
Comment (NisargIO) [inline packages/core/src/checks/expo/utils/read-expo-app-config.ts]: slop

File: packages/core/src/checks/expo/utils/read-expo-app-config.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  try {
+    return fs.readFileSync(filePath, "utf-8");
+  } catch {
+    return null;
+  }
+};
+
+const unwrapExpoConfig = (parsed: unknown): Record<string, unknown> | null => {
+  if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) return null;
```
Comment (NisargIO) [inline packages/core/src/checks/expo/utils/read-expo-app-config.ts]: zod

File: packages/core/src/checks/expo/utils/read-expo-app-config.ts (hunk 3 above)
Comment (rayhanadev) [reply packages/core/src/checks/expo/utils/read-expo-app-config.ts]: Done in `d648b03c` — dropped the `readFileSafe` wrapper (the read is now an inline `JSON.parse(fs.readFileSync(...))` in `decodeExpoConfig`), and removed the `getNestedConfigValue` walker in favor of typed field access. The whole file is ~40% smaller.

File: packages/core/src/checks/expo/utils/read-expo-app-config.ts (hunk 4 above)
Comment (rayhanadev) [reply packages/core/src/checks/expo/utils/read-expo-app-config.ts]: Done in `d648b03c` — replaced the hand-rolled `unwrapExpoConfig` (typeof / Array.isArray) with an `effect/Schema` struct decode (`Schema.Struct` + `Option.getOrNull(Schema.decodeUnknownOption(...))`). Used effect/Schema rather than zod because that's the schema lib core already uses (schemas.ts, calculate-score.ts) — core has no zod dependency. Checks now read typed `config.newArchEnabled` / `config.updates.disableAntiBrickingMeasures`. Lmk if you'd prefer an actual zod dep instead.

File: packages/core/src/checks/expo/check-reanimated-new-arch.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  if (appConfig.config?.newArchEnabled !== false) return [];
+
+  return [
+    buildExpoDiagnostic({
+      rule: "expo-reanimated-v4-requires-new-arch",
+      filePath: appConfig.configFile ?? "app.json",
+      message:
+        "react-native-reanimated v4 supports only the New Architecture, but `newArchEnabled: false` is set in your app config, so the app will crash on first launch.",
+      help: "Remove `newArchEnabled: false` from your app config (the New Architecture is the default on SDK 52+), or pin react-native-reanimated to v3 if you must stay on the legacy architecture.",
+    }),
```
[the remark this answers is not in the document]
Comment (rayhanadev) [reply packages/core/src/checks/expo/check-reanimated-new-arch.ts:43]: Good catch — fixed in `9a5a4b70`. All three now emit `error` (not the `buildExpoDiagnostic`/inline `warning` default), so a guaranteed first-launch crash (`expo-reanimated-v4-requires-new-arch`), an app-bricking config (`expo-updates-no-unsafe-production-config`), and a broken Metro/Babel transform (`rn-no-metro-babel-preset`) surface in default scans rather than hiding behind `--warnings`. Added severity assertions to lock it in. The other expo-doctor-ported checks (lockfile, gitignore, metro-config, …) stay `warning` since they're not catastrophic.
